filesetup.c

```diff
@@ -833,12 +833,26 @@ static unsigned long long get_fs_free_counts(struct thread_data *td)
 uint64_t get_start_offset(struct thread_data *td, struct fio_file *f)
 {
 	struct thread_options *o = &td->o;
+	unsigned long long offset;
 
 	if (o->file_append && f->filetype == FIO_TYPE_FILE)
 		return f->real_file_size;
 
-	return td->o.start_offset +
-		td->subjob_number * td->o.offset_increment;
+	if (o->start_offset_percent > 0) {
+		// calculate the raw offset
+		offset = (f->real_file_size * o->start_offset_percent / 100) +
+			(td->subjob_number * td->o.offset_increment);
+
+		// block align the offset at the next available boundary
+		offset = (offset / o->ba[o->td_ddir] +
+			(offset % o->ba[o->td_ddir] != 0))  // ceiling(offset/blockalignment)
```

Comment:
Yeah, o->td_ddir is not the way to go. You want something ala:

`if (fio_option_is_set(o, ba)) {
    align_bs = min(td->o.ba[DDIR_READ], td->o.ba[DDIR_WRITE]);
    align_bs = min(td->o.ba[DDIR_TRIM], align_bs);
} else
    align_bs = td_min_bs(td);
`